// ipx/dist/index.cjs
'use strict';

const nodeFs = require('./shared/ipx.ebaf2d0c.cjs');
require('defu');
require('ufo');
require('h3');
require('image-meta');
require('destr');
require('@fastify/accept-negotiator');
require('etag');
require('ofetch');
require('pathe');

function unstorageToIPXStorage(storage, prefix) {
  const resolveKey = (id) => prefix + ":" + id;
  return {
    name: "ipx:" + (storage.name || "unstorage"),
    async getMeta(id, opts = {}) {
      if (!storage.getMeta) {
        return;
      }
      const storageKey = resolveKey(id);
      const meta = await storage.getMeta(storageKey, opts);
      return meta;
    },
    async getData(id, opts = {}) {
      if (!storage.getItemRaw) {
        return;
      }
      const storageKey = resolveKey(id);
      const data = await storage.getItemRaw(storageKey, opts);
      return data;
    }
  };
}

exports.createIPX = nodeFs.createIPX;
exports.createIPXH3App = nodeFs.createIPXH3App;
exports.createIPXH3Handler = nodeFs.createIPXH3Handler;
exports.createIPXNodeServer = nodeFs.createIPXNodeServer;
exports.createIPXPlainServer = nodeFs.createIPXPlainServer;
exports.createIPXWebServer = nodeFs.createIPXWebServer;
exports.ipxFSStorage = nodeFs.ipxFSStorage;
exports.ipxHttpStorage = nodeFs.ipxHttpStorage;
exports.unstorageToIPXStorage = unstorageToIPXStorage;
